infrastructure/src/main/java/org/corfudb/infrastructure/log/FileSystemAgent.java

```diff
@@ -0,0 +1,132 @@
+        long logSizeLimit = getLogSizeLimit();
+
+        logSizeQuota = new ResourceQuota("LogSizeQuota", logSizeLimit);
+        logSizeQuota.consume(initialLogSize);
+        log.info("StreamLogFiles: {} size is {} bytes, limit {}", config.logDir, initialLogSize, logSizeLimit);
+    }
+
+    private long getLogSizeLimit() {
+        long fileSystemCapacity = getFileSystemCapacity();
+
+        // Derived size in bytes that normal writes to the log unit are capped at.
+        // This is derived as a percentage of the log's filesystem capacity.
+        return (long) (fileSystemCapacity * config.limitPercentage / 100);
+    }
+
+    public static void init(ResourceQuotaConfig config) {
+        instance = Optional.of(new FileSystemAgent(config));
+    }
+
+    public static boolean configured(){
+        return instance.isPresent();
+    }
+
+    public static ResourceQuota getResourceQuota() {
+        Supplier<IllegalStateException> err = () -> new IllegalStateException("ResourceQuota not configured");
+        return instance.orElseThrow(err).logSizeQuota;
+    }
+
+    /**
+     * Estimate the size (in bytes) of a directory.
+     * From https://stackoverflow.com/a/19869323
+     */
+    @VisibleForTesting
+    long estimateSize() {
+        final AtomicLong size = new AtomicLong(0);
+        try {
+            Files.walkFileTree(config.logDir, new SimpleFileVisitor<Path>() {
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+                    size.addAndGet(attrs.size());
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFileFailed(Path file, IOException exc) {
+                    // Skip folders that can't be traversed
+                    log.error("skipped: {}", file, exc);
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+
+            return size.get();
+        } catch (IOException ioe) {
+            throw new IllegalStateException(ioe);
+        }
+    }
+
+    /**
+     * Get corfu log dir partition size
+     *
+     * @return total capacity of the file system that owns the log files.
+     */
+    private long getFileSystemCapacity() {
+        Path corfuDir = config.logDir.getParent();
+        try {
+            return Files.getFileStore(corfuDir).getTotalSpace();
+        } catch (IOException e) {
+            throw new IllegalStateException("Failed reading corfu log directory, path: " + corfuDir, e);
+        }
+    }
+
+    public static class ResourceQuotaConfig {
+        private final Path logDir;
+        private final double limitPercentage;
+
+        public ResourceQuotaConfig(ServerContext serverContext) {
+            String limitParam = serverContext.getServerConfig(String.class, "--log-size-quota-percentage");
+            limitPercentage = Double.parseDouble(limitParam);
```

Comment:
This will cause a NPE if `--log-size-quota-percentage` isn't given as a command line parameter. It doesn't seem to be populated with a default value otherwise.

Reply:
There is a default (it's 100%) value https://github.com/CorfuDB/CorfuDB/blob/4dad5fdc5a5534b3d8b96abc7372b63bb5acaeb3/infrastructure/src/main/java/org/corfudb/infrastructure/CorfuServer.java#L164